klc-cert-manager/controllers/keptnwebhookcontroller/certs.go

```diff
@@ -0,0 +1,226 @@
+		return true
+	} else if cs.rootPublicCert, err = x509.ParseCertificate(block.Bytes); err != nil {
+		//log.Info("failed to parse root certificates, renewing", "error", err)
+		return true
+	} else if now.After(cs.rootPublicCert.NotAfter.Add(-renewalThreshold)) {
+		//log.Info("root certificates are about to expire, renewing", "current", now, "expiration", cs.rootPublicCert.NotAfter)
+		return true
+	}
+
+	if block, _ := pem.Decode(cs.Data[RootKey]); block == nil {
+		//log.Info("failed to parse root key, renewing", "error", "can't decode PEM file")
+		return true
+	} else if cs.rootPrivateKey, err = x509.ParseECPrivateKey(block.Bytes); err != nil {
+		//log.Info("failed to parse root key, renewing", "error", err)
+		return true
+	}
+
+	return false
+}
+
+func (cs *Certs) validateServerCerts(now time.Time) bool {
+	if cs.Data[ServerKey] == nil || cs.Data[ServerCert] == nil {
+		//log.Info("no server certificates found, creating")
+		return true
+	}
+
+	isValid, err := kubeutils.ValidateCertificateExpiration(cs.Data[ServerCert], renewalThreshold, now)
+	if err != nil || !isValid {
+		//log.Info("server certificate failed to parse or is outdated")
+		return true
+	}
+	return false
+}
+
+func (cs *Certs) generateRootCerts(domain string, now time.Time) error {
+	// Generate CA root keys
+	//log.Info("generating root certificate")
+	privateKey, err := cs.generatePrivateKey(RootKey)
+	if err != nil {
+		return err
+	}
+	cs.rootPrivateKey = privateKey
+
+	// Generate CA root certificate
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return fmt.Errorf("failed to generate serial number for root certificate: %w", err)
+	}
+
+	cs.rootPublicCert = &x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Country:            []string{"AT"},
+			Province:           []string{"UA"},
+			Locality:           []string{"Linz"},
+			Organization:       []string{"Dynatrace"},
+			OrganizationalUnit: []string{"OneAgent Webhook"},
+			CommonName:         domain,
+		},
+		IsCA: true,
+
+		NotBefore: now,
+		NotAfter:  now.Add(365 * 24 * time.Hour),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	rootPublicCertDER, err := x509.CreateCertificate(
+		rand.Reader,
+		cs.rootPublicCert,
+		cs.rootPublicCert,
+		cs.rootPrivateKey.Public(),
+		cs.rootPrivateKey)
+	if err != nil {
+		return fmt.Errorf("failed to generate root certificate: %w", err)
+	}
+
+	cs.Data[RootCertOld] = cs.Data[RootCert]
+	cs.Data[RootCert] = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: rootPublicCertDER})
+
+	//log.Info("root certificate generated")
+	return nil
+}
+
+func (cs *Certs) generateServerCerts(domain string, now time.Time) error {
+	// Generate server keys
+	//log.Info("generating server certificate")
```

Comment:
should we uncomment all the commented out log statements? otherwise, let's remove them if we do not need them anymore